src/MemoryRouter.tsx

```diff
@@ -145,9 +146,20 @@ export class MemoryRouter extends BaseRouter {
     const shallow = options?.shallow || false;
     const pathname = removeTrailingSlash(urlObject.pathname || "");
     const query = urlObject.query || {};
-    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, urlObject.hash);
+    const hash = urlObject.hash;
+    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, hash);
 
-    this.events.emit("routeChangeStart", asPath, { shallow });
+    const isHashChange = this.hash !== hash;
+    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
+    const isRouteChange = isQueryChange || this.pathname !== pathname;
+
+    if (isHashChange) {
+      this.events.emit("hashChangeStart", asPath, { shallow });
+    }
+
+    if (isRouteChange) {
+      this.events.emit("routeChangeStart", asPath, { shallow });
```

Comment:
In a Next app, if I was to call `router.push(window.location)` (ie. the path doesn't change at all), any idea if it would still fire the `routeChangeStart/Complete` events?

Reply:
yes you can test it here: https://codesandbox.io/s/nextjs-routerevent-tests-fwtzu